### PR TITLE
Move CMFD to Python through C API - Replace Fortran, update docs and test suite

### PR DESCRIPTION
**Pull request by shikhar413**
This PR moves CMFD implementation currently written in Fortran to Python through the C API. CMFD cross sections and matrices are populated using a vectorized Numpy approach, and these arrays are passed into the solver which is written in C++. More implementation and timing details can be found in #1103. Support for outputting CMFD variables to statepoint and plotting CMFD meshes has not been implemented yet and will be added to a future PR.

This PR also goes ahead and updates the documentation and test suite accordingly.
- With regards to documentation, I wasn't sure where to include all the detail found in `openmc/docs/source/io_formats/cmfd.rst`, since xml files will no longer be used to call CMFD. I've outlined the Python functions available to users in order to set CMFD parameters in `docs/source/pythonapi/base.rst`
- `CMFDTestingHarness` has been modified so that additional CMFD parameters used for comparison are passed into `__init__` in memory and appended to results such as k-effective and tally results that are extracted from statepoint file. I'm thinking I should create accessor functions that would return variables such as `cmfd_run._indices` and `cmfd_run._k_cmfd` that are currently private variables?
- The changes to `tests/regression_tests/cmfd_feed/results_true.dat` are due to the fact that batchwise CMFD results are only stored during batches where CMFD is turned on.  In addition, there are slight discrepancies in "cmfd balance" because the Fortran version relies on the k-effective value at a specific batch, while the C API k-effective returns a minimum variance estimate when there are enough realizations in the active cycles.

### CODE REVIEW

File: openmc/cmfd.py [hunk 1]
```diff
@@ -12,12 +12,65 @@
 
 from collections.abc import Iterable
 from numbers import Real, Integral
-from xml.etree import ElementTree as ET
 import sys
-
```
Comment (paulromano) [inline openmc/cmfd.py]: Keep a blank line between standard-library imports and third-party library imports

File: openmc/cmfd.py [hunk 2]
```diff
@@ -12,12 +12,65 @@
...
+import time
+from ctypes import c_int, c_double
+# See if mpi4py module can be imported, define have_mpi global variable
+try:
+    from mpi4py import MPI
+    have_mpi = True
+except ImportError:
+    have_mpi = False
+
+import openmc.capi
```
Comment (paulromano) [inline openmc/cmfd.py]: We've been trying to avoid having a top-level `import openmc.capi` because it requires anyone who just wants to `import openmc` to have libopenmc set up in the right place. I would suggest moving the declarations of the functions from the C API (restype, argtypes, etc.) to `openmc/capi/core.py` and then have `from openmc.capi import _dll` in any function here where the C API is used.

File: openmc/cmfd.py [hunk 3]
```diff
@@ -12,12 +12,65 @@
...
+array_1d_int = npct.ndpointer(dtype=np.int32, ndim=1, flags='CONTIGUOUS')
+array_1d_dble = npct.ndpointer(dtype=np.double, ndim=1, flags='CONTIGUOUS')
+
+# Setup the return types and argument types for C++ functions
+openmc.capi._dll.openmc_initialize_linsolver.restype = None
+openmc.capi._dll.openmc_initialize_linsolver.argtypes = [array_1d_int, c_int,
+    array_1d_int, c_int, c_int, c_double, array_1d_int, array_1d_int]
+openmc.capi._dll.openmc_run_linsolver.restype = c_int
+openmc.capi._dll.openmc_run_linsolver.argtypes = [array_1d_dble, array_1d_dble,
+    array_1d_dble, c_double]
```
Comment (paulromano) [inline openmc/cmfd.py]: Move this chunk of code to `openmc/capi/core.py`

File: openmc/cmfd.py [hunk 4]
```diff
@@ -12,12 +12,65 @@
...
+"""
+# Maximum/minimum neutron energies
+_ENERGY_MAX_NEUTRON = np.inf
+_ENERGY_MIN_NEUTRON = 0.
+
+# Tolerance for detecting zero flux values
+_TINY_BIT = 1.e-8
+
+# For non-accelerated regions on coarse mesh overlay
+_CMFD_NOACCEL = 99999
```
Comment (paulromano) [inline openmc/cmfd.py]: Would a value like -1 work just as well?

File: openmc/cmfd.py [hunk 5]
```diff
@@ -12,12 +12,65 @@
...
+
+# Constant to represent a zero flux "albedo"
+_ZERO_FLUX = 999.0
+
+# Map that returns index of current direction in numpy current matrix
+_CURRENTS = {
+    'out_left'  : 0, 'in_left'  : 1, 'out_right':  2, 'in_right': 3,
```
Comment (paulromano) [inline openmc/cmfd.py]: PEP8: No spaces before `:`

File: openmc/cmfd.py [hunk 6]
```diff
@@ -177,115 +235,364 @@ def _get_xml_element(self):
...
+        listed in the following order: [-x +x -y +y -z +z]. Set to
+        [1, 1, 1, 1, 1, 1] if not specified by user.
+    coremap : numpy.ndarray
+        Coremap for coarse mesh overlay, defining acceleration regions in CMFD
+        problem, created from map variable in CMFD class. Each accelerated
+        region is given a unique index to map each spatial mesh to a
+        corresponding row within the CMFD matrices. Non-accelerated regions are
+        set with value _CMFD_NOACCEL. If a user does not specify a CMFD map,
+        coremap is set to treat each spatial mesh as an accelerated region.
+        For non-accelerated regions on coarse mesh overlay
```
Comment (paulromano) [inline openmc/cmfd.py]: Something's missing here

File: docs/source/pythonapi/base.rst [hunk 7]
```diff
@@ -208,3 +197,53 @@ Various classes may be created when performing tally slicing and/or arithmetic:
...
+   openmc.CMFDRun.norm
+   openmc.CMFDRun.cmfd_adjoint_type
+   openmc.CMFDRun.cmfd_power_monitor
+   openmc.CMFDRun.cmfd_run_adjoint
+   openmc.CMFDRun.cmfd_shift
+   openmc.CMFDRun.cmfd_stol
+   openmc.CMFDRun.cmfd_spectral
+   openmc.CMFDRun.cmfd_reset
+   openmc.CMFDRun.cmfd_write_matrices
+   openmc.CMFDRun.gauss_seidel_tolerance
```
Comment (paulromano) [inline docs/source/pythonapi/base.rst]: All of these are properties, so they won't show up correctly when listed as functions. In fact, there's no reason to list them explicitly: the autogenerated documentation for the `CMFDRun` class will show all the properties.

File: openmc/cmfd.py [hunk 8]
```diff
@@ -147,9 +204,10 @@ def albedo(self, albedo):
     def map(self, meshmap):
         check_type('CMFD mesh map', meshmap, Iterable, Integral)
         for m in meshmap:
-            check_value('CMFD mesh map', m, [1, 2])
+            check_value('CMFD mesh map', m, [0, 1])
         self._map = meshmap
 
+    # REMOVE this method
```
Comment (paulromano) [inline openmc/cmfd.py]: Follow your advice? :)

File: openmc/cmfd.py [hunk 9]
```diff
@@ -296,225 +603,2224 @@ def norm(self):
...
-        """Create a cmfd.xml file using the class data that can be used for an OpenMC
-        simulation.
+    @gauss_seidel_tolerance.setter
+    def gauss_seidel_tolerance(self, gauss_seidel_tolerance):
+        check_type('CMFD Gauss-Seidel tolerance', gauss_seidel_tolerance,
+                   Iterable, Real)
+        check_length('Gauss-Seidel tolerance', gauss_seidel_tolerance, 2)
+        self._gauss_seidel_tolerance = gauss_seidel_tolerance
+
+    def run(self, omp_num_threads=None, intracomm=None):
```
Comment (paulromano) [inline openmc/cmfd.py]: Rather than having arguments for `omp_num_threads` and `intracomm`, I would just have it accept `**kwargs` that gets passed to `openmc.capi.init` -- this enables a user to pass whatever arguments (and intracommunicator) they want.

File: openmc/cmfd.py [hunk 10]
```diff
@@ -296,225 +603,2224 @@ def norm(self):
...
+    def _write_matrix(self, matrix, base_filename):
+        """Write a numpy matrix to file and also save it in .npz format. This
+        particular format allows users to load the variable directly in a Python
+        session with scipy.sparse.load_npz()
+
+        Parameters
+        ----------
+        matrix : scipy.sparse.spmatrix
+            Sparse matrix that will be saved
+        base_filename : string
```
Comment (paulromano) [inline openmc/cmfd.py]: `base_filename : str`

File: openmc/cmfd.py [hunk 11]
```diff
@@ -296,225 +603,2224 @@ def norm(self):
...
+        # Write row, col, and data of each entry in sparse matrix. This ignores
+        # all zero-entries, and indices are written with zero-based indexing
+        with open(base_filename+'.dat', 'w') as fh:
+            for row in range(matrix.shape[0]):
+                # Get all cols for particular row in matrix
+                cols = matrix.indices[matrix.indptr[row]:matrix.indptr[row+1]]
+                # Get all data entries for particular row in matrix
+                data = matrix.data[matrix.indptr[row]:matrix.indptr[row+1]]
+                for i in range(len(cols)):
+                    fh.write('({0:3d}, {1:3d}): {2:0.8f}\n'.format(
```
Comment (paulromano) [inline openmc/cmfd.py]: The following would suffice:
```Python
fh.write('({:3d}, {:3d}): {:0.8f}\n'.format(
```

File: openmc/cmfd.py (hunk 9 above)
Comment (paulromano) [reply openmc/cmfd.py]: Actually, I see that you explicitly need the intracomm because it's used directly. Nevertheless, things like the number of threads could be picked up by `**kwargs`

File: openmc/cmfd.py [hunk 12]
```diff
@@ -296,225 +603,2224 @@ def norm(self):
...
+                        sourcecounts, where=divide_condition, \
+                        out=np.ones_like(self._cmfd_src))
+
+            if not self._cmfd_feedback:
+                return
+
+            # Broadcast weight factors to all procs
+            if have_mpi:
+                self._weightfactors = self._intracomm.bcast(
+                                      self._weightfactors, root=0)
```
Comment (paulromano) [inline openmc/cmfd.py]: The default root is zero, so that argument can be omitted.

File: openmc/cmfd.py [hunk 13]
```diff
@@ -296,225 +603,2224 @@ def norm(self):
...
-        simulation.
+    @gauss_seidel_tolerance.setter
+    def gauss_seidel_tolerance(self, gauss_seidel_tolerance):
+        check_type('CMFD Gauss-Seidel tolerance', gauss_seidel_tolerance,
+                   Iterable, Real)
+        check_length('Gauss-Seidel tolerance', gauss_seidel_tolerance, 2)
+        self._gauss_seidel_tolerance = gauss_seidel_tolerance
+
+    def run(self, omp_num_threads=None, intracomm=None):
+        """Public method to run OpenMC with CMFD
```
Comment (paulromano) [inline openmc/cmfd.py]: Suggest rewording to "Run OpenMC with coarse mesh finite difference acceleration"

**Comment by paulromano** [comment]
I just took a look at the [coverage](https://coveralls.io/builds/20271929/source?filename=openmc%2Fcmfd.py) of cmfd.py and it shows a few more things that are not being tested:
- Writing matrices/vectors to disk
- Power monitor
- CMFD tally reset
- Adjoint
- Effective downscatter (also noticed that the cmfd_downscatter property is not documented)

Unrelated to that, I'd recommend running [pycodestyle](https://github.com/PyCQA/pycodestyle) on cmfd.py to see where you can improve the style to better match PEP8. I wouldn't pay too much attention to its complaints about spaces after commas (anywhere you have an expression like `array[:,:,:]` it will complain that it should be `array[:, :, :]` but I wouldn't worry about that).

**Comment by shikhar413** [comment]
Thanks for the comments. Just a heads up it'll take me some time to get to these changes right now because of the ExaSMR meeting but I will address these issues after the meeting.

**Comment by paulromano** [comment]
@shikhar413 There's a new merge conflict that you'll need to resolve.

**Comment by shikhar413** [comment]
Apologies for the delay in getting this done. I have addressed most of the comments but while updating the test suite to increase code coverage, I've been running into seg faults in `nuclide_clear` when calling `free_memory_nuclide`. I'll need a bit more time to figure out what the issue is.

**Comment by paulromano** [comment]
@shikhar413 Did you merge the most recent develop branch into your own branch?

**Comment by shikhar413** [comment]
Never mind, the bug was completely my own doing. Updating the test suite right now

**Comment by shikhar413** [comment]
All comments should be addressed now. A few notes:

- I removed most variables in the CMFDRun docstring that users cannot modify and included only the ones that are related to CMFD input parameters, CMFD results (such as k_cmfd, entropy etc), and variables used in the test suite (indices and cmfd_src). Let me know if I should include other CMFD-related variables such as `flux`, `totalxs`, `dhat` etc. in the docstring or elsewhere to make clear what these variables are.
- Right now in `openmc/capi/core.py`, the cmfd_solver functions do not have a errcheck variable defined, is this necessary?
- All variables used to compare results in the test suite now have public accessor functions in CMFDRun (i.e. `k_cmfd`, `dom`, `entropy`, `balance`, `src_cmp`, `cmfd_src`, `indices`)
- `display` is now a dictionary where users can display which output metrics they want to see, and is set as `cmfd_run.display = {'entropy': True}`. I couldn't figure out a way to set this directly such that users can call `cmfd_run.output['dominance_ratio'] = True` while still retaining error checking on the key-value entries.
- In order to get around not importing openmc.capi in `cmfd.py` I import `_dll` and all modules within openmc.capi used for the CMFD algorithm
- The 2-group and n-group solvers work on a 2x2 lattice of 4 BEAVRS pincells surrounded by water, with height 2cm. The CMFD algorithm is very sensitive to number of groups and group boundaries so the n-group problem has three groups which is not realistic but tests the pertinent code. Adding more groups results in scenarios where there isn't an accurate flux within each group and the GS solver does not work.

**Comment by shikhar413** [comment]
One more thing:

- Regarding test coverage for `power_monitor`, the results from setting this variable to True are outputted to the buffer and I wasn't sure how to extract this output as a test case. Also, `tally_reset` is tested in the n group solver, and `downscatter` is tested in the 2 group solver

**Comment by shikhar413** [comment]
Was running into some issues with the test suite but they should be resolved now. I had to free memory for the arrays created in C++ and change the import statements in `cmfd.py`. Initially I included the statement `from openmc.capi import _dll, core, settings, filter, mesh, tally` to get around not importing `openmc.capi` entirely, but this would import the C API version of the `filter` module, causing the test suite to fail when it would expect the Python API version of `filter`. 

Now, I included the `import openmc.capi` statement back into `cmfd.py`, but `cmfd.py` is no longer imported when importing openmc, since the current CMFD implementation has a clear dependence on the C API and is not useful without it. `from openmc import cmfd` needs to be included in all files running CMFD and also requires that openmc.capi be able to be imported. I can go ahead and update the documentation to include this if this is the way to go. Turns out that importing cmfd.py in openmc/__init__.py also imports the functions in the `openmc/checkvalue.py` module that is required by some of the regression tests in the test suite, so now all functions in `openmc/checkvalue.py` are explicitly imported in `openmc/__init__.py`

File: openmc/cmfd.py [hunk 14]
```diff
@@ -383,138 +580,2177 @@ def run_adjoint(self, run_adjoint):
...
+            self._intracomm = intracomm
+        elif intracomm is None and have_mpi:
+            self._intracomm = MPI.COMM_WORLD
+
+        # Check keyword arguments and pass to C API init function
+        if 'openmc_args' in kwargs:
+            openmc.capi.init(args=kwargs['openmc_args'],
+                             intracomm=self._intracomm)
+        else:
+            openmc.capi.init(intracomm=self._intracomm)
```
[the remark this answers is not in the document]
Comment (shikhar413) [reply openmc/cmfd.py]: Neat!

**Comment by paulromano** [comment]
I'm happy with this when @smharper's requests have been resolved. Thanks @shikhar413!

File: openmc/cmfd.py [hunk 15]
```diff
@@ -383,138 +576,2170 @@ def run_adjoint(self, run_adjoint):
...
+
+        # Start timer for build
+        time_start_buildcmfd = time.time()
+
+        # Build loss and production matrices
+        loss, prod = self._build_matrices(physical_adjoint)
+
+        # Check for mathematical adjoint calculation
+        if adjoint and self._adjoint_type == 'math':
+            loss, prod = self._compute_adjoint(loss, prod)
```
[the remark this answers is not in the document]
Comment (shikhar413) [reply openmc/cmfd.py]: This is much cleaner, thanks!

**Comment by shikhar413** [comment]
I think I've addressed everything now, @smharper care to look over the `CMFDRun.run` function once more and let me know if `openmc.capi.run_in_memory` is used as you were envisioning?

File: openmc/cmfd.py [hunk 16]
```diff
@@ -608,73 +608,69 @@ def gauss_seidel_tolerance(self, gauss_seidel_tolerance):
...
+            All keyword arguments are passed to
+            :func:`openmc.capi.run_in_memory`.
 
         """
         # Store intracomm for part of CMFD routine where MPI reduce and
         # broadcast calls are made
-        if intracomm is not None:
-            self._intracomm = intracomm
-        elif intracomm is None and have_mpi:
+        if 'intracomm' in kwargs.keys() and kwargs['intracomm'] is not None:
```
Comment (smharper) [inline openmc/cmfd.py]:
```python
          if 'intracomm' in kwargs and kwargs['intracomm'] is not None: 
```
The `.keys()` is not needed.

**Comment by shikhar413** [comment]
No problem, thanks for the thorough review @paulromano and @smharper. This was a huge learning experience for me as well!
